tests/smoke_tests/test_basic.py

```diff
@@ -659,6 +659,68 @@ def test_kubernetes_context_failover():
         smoke_tests_utils.run_one_test(test)
 
 
+def test_launch_and_exec_async(generic_cloud: str):
+    """Test if the launch and exec commands work correctly with --async."""
+    name = smoke_tests_utils.get_cluster_name()
+    to_down = smoke_tests_utils.get_cluster_name() + '-down'
+    test = smoke_tests_utils.Test(
+        'launch_and_exec_async',
+        [
+            f'sky launch -c {name} -y --async',
+            # Async exec.
+            f'sky exec {name} echo --async',
+            # Async exec and cancel immediately.
+            (f's=$(sky exec {name} echo --async) && '
+             'echo "$s" && '
+             'cancel_cmd=$(echo "$s" | grep "To cancel the request" | '
+             'sed -E "s/.*run: (sky api cancel .*).*/\\1/") && '
+             'echo "Extracted cancel command: $cancel_cmd" && '
+             '$cancel_cmd'),
+            # Sync exec.
+            f'sky exec {name} echo',
+            # Cluster must be UP since the sync exec has been completed.
+            f'sky status {name} | grep "UP"',
+            # The async one might be scheduled later and the job ID is
+            # non-deterministic, so we wait for both jobs asynchronously.
+            smoke_tests_utils.RetriableCommand(
+                f'sky logs {name} 1 --status | grep "SUCCEEDED"',
+                retry_interval=5,
+                max_attempts=5,
+            ),
```

Comment:
This should not be retried, as we should make sure after `sky exec` the jobs are succeeded.

Reply:
Yes, but the ID (job submission order) is non-deterministic here. It is possible to get the job ID by tailing the log of the async request, but since variables cannot be transferred between commands for now, I just wait for both jobs to succeed for simplicity